src/main/scala/piconot/semantics/Interpreter.scala

```diff
@@ -0,0 +1,115 @@
+package piconot
+
+/**
+ * @author mvalentine
+ */
+
+import piconot.ir._
+import picolib.semantics._
+import java.io.File
+import scalafx.application.JFXApp
+
+import picolib.maze.Maze
+
+class interpreter extends JFXApp {
+  def picobot(mazename: String)(rs: Seq[Rule]*): List[Rule] = {
+    val rules = rs.flatten.toList
+    
+    val maze = Maze("resources" + File.separator + s"${mazename}.txt")
+    object RuleBot extends Picobot(maze, rules)
+      with TextDisplay with GUIDisplay
+    RuleBot.run()
+    stage = RuleBot.mainStage
+    
+    rules
+  }
+}
+
+package object semantics extends interpreter{
+  
+  case class RulesDontMakeSense(msg: String) extends Exception(msg)
+def eval(ast: AST)(mazeName: String): List[Rule] = {
+    val rules = evalTransformers(ast)
+    println(rules)
+    picobot(mazeName)(rules)
+    rules.toList
+  }
+  
+  def evalTransformers(ast: AST): Seq[Rule] = {
+    val DefaultSurroundings: Surroundings = Surroundings(Anything, Anything, Anything, Anything)
+  
+    val UndefinedStateString: String = "!!__ UN DE FI NE D __!!"
+    val UndefinedState: State = new State(UndefinedStateString)
+    
+    val DefaultRule: Rule = Rule(UndefinedState, 
+                            DefaultSurroundings, 
+                            StayHere, 
+                            UndefinedState)
+    ast match {
+      case BaseTransformer(aug: Augment) => {
+        Seq(applyAugment(aug, DefaultRule))
+      }
+      case AugmentTransformer(aug: Augment, trans: Transformers) =>
+          evalTransformers(trans).map {rule => applyAugment(aug, rule)}
+          
+      case BaseTransformers(trans: Transformer) => evalTransformers(trans)
+      case BracedTransformers(multiTrans: MultiTransformers) => evalTransformers(multiTrans)
+      
+      case SingleMultiTransformers(t: Transformer) => evalTransformers(t)
+      case MutlipleMultiTransformers(transL, multiT) => 
+        evalTransformers(transL) ++ evalTransformers(multiT)
+      case ElseTransformerBasic(Move(dir), trans) => {
+        evalTransformers(BaseTransformer(Move(dir))) ++ evalTransformers(AugmentTransformer(Restrict(dir, Blocked), trans))
+      }
+      case ElseTransformerComplex(Move(dir), trans1, trans2) => {
+        evalTransformers(AugmentTransformer(Move(dir), trans1)) ++ evalTransformers(AugmentTransformer(Restrict(dir, Blocked), trans2))
+      }
+    }
+  }
```

Comment:
i think its so cool how you can just throw match statements into the middle of a function, i think in SML you can only match on the args